shivyc/tree/nodes.py

```diff
@@ -542,42 +544,56 @@ def get_storage(self, spec_kinds, spec_range):
 
         return storage
 
-    def parse_struct_spec(self, node, redec, symbol_table):
-        """Parse a struct ctype from the given decl_nodes.Struct node.
+    def parse_struct_union_spec(self, node, redec, symbol_table):
+        """Parse struct or union ctype from the given decl_nodes.Struct node.
 
-        node (decl_nodes.Struct) - the Struct node to parse
+        node (decl_nodes.Struct) - the Struct or Union node to parse
         redec (bool) - Whether this declaration is alone like so:
 
            struct S;
+           union U;
 
         or declares variables/has storage specifiers:
 
            struct S *p;
            extern struct S;
+           union U *u;
+           extern union U;
 
         If it's the first, then this is always a forward declaration for a
         new `struct S` but if it's the second and a `struct S` already
         exists in higher scope, it's just using the higher scope struct.
         """
         has_members = node.members is not None
+        kind = node.kind
```

Comment:
It's a bit cumbersome to have to check `kind` so often in the body of this function. It might be easier to check kind here, like so
```python
if node.kind == token_kinds.struct_kw:
  symbol_table_lookup = symbol_table.lookup_struct
  this_ctype = StructCType
else:
  (analogous to above)
```
Then, every time below that you call `symbol_table.lookup_struct`, you could just use `symbol_table_lookup`, and similarly for the other variables.